### ISSUE TITLE
Can I add fields to a CSV import template?

### ISSUE THREAD

**Issue by megandalster** [farmOS/farmOS#199]
I'm trying to import Feed Purchases from FarmData into my FarmOS install. Your CSV template accommodates only a few fields, but my download from FarmData has a lot more -- for example, vendor, unit, unit_price, unit_weight. Is there a way in FarmOS to modify the CSV upload template to accommodate these additional fields? I see these fields in the headers on your Purchases page, so I'd like the import to populate all those columns.

Thanks for your suggestion on issue farmOS/farmOS#198. I will start a conversation about FarmData on your community forum as soon as I have a little more confidence with the data export-import process from FarmData.

**Comment by mstenta** [comment]
(Transferred this issue to the main farmOS issue queue, since it is not directly related to farmOS.org)

**Comment by megandalster** [comment]
You're right, input logs don't quite do the job, though somewhere in your documentation suggested that they might.

Anyway, here are the columns exported from AnimalData (a companion app to FarmData, both developed at Dickinson), along with a few rows of real data.  

id | purch_date | type | subtype | animal_group | vendor | unit | purchased | price_unit | weight_unit | comments

22 | 2016-07-08 | CHICKEN FEED | LAYER MIX | LAYING HENS | JASON FRITZ | BAG | 6.00 | 30.00 | 80.00 |  
41 | 2016-08-05 | CHICKEN FEED | LAYER MIX | LAYING HENS | JASON FRITZ | BAG | 7.00 | 30.00 | 80.00 |  
43 | 2016-08-05 | MINERALS | SHEEP NUTRI BALANCER | SHEEP | JASON FRITZ | BAG | 1.00 | 45.00 | 50.00 |  
44 | 2016-08-05 | MINERALS | GRAZIERS CHOICE | CATTLE | JASON FRITZ | BAG | 1.00 | 39.50 | 50.00 |  
39 | 2016-08-26 | BEDDING | STRAW - SMALL SQUARES | CATTLE | JASON FRITZ | SMALL SQUARE BALE | 1.00 | 56.00 | 1.00 | Delivery fee for 68 bales
40 | 2016-08-26 | BEDDING | STRAW - SMALL SQUARES | CATTLE | JASON FRITZ | SMALL SQUARE BALE | 68.00 | 4.00 | 30.00 |  

It looks like the Purchase Log type in FarmOS will accommodate these columns, but its associated importer is missing many of the needed fields.

My overall goal is to merge Dickinson's FarmData and AnimalData installs into a new install called FarmData2, which will hopefully become a new FarmOS instance.  The trick is to be able to import all/most of the data from these two older apps into FarmData2.   The importers are crucial for making this work.

I'm working closely with Matt Steiman on all of this.  

Many thanks for any help you can provide.

**Comment by mstenta** [comment]
Thanks! Very helpful to see.

Two things:

1. We should add mapping targets to the farmOS Purchase log importer for the fields that *are* currently on Purchase logs.  More on that below...
2. You will need to decide what to do with data that does *not* have a destination in farmOS currently.

Re: 1, taking a look at the `farm_ledger` module, I see we are not adding the importer mappings for purchase logs in the same way that we are for sale logs.  That's easy enough to fix.  That was an oversight when Purchase logs were added.  I'll make a commit shortly...

Re: 2, some thoughts:

- `id` - Is this referenced elsewhere in AnimalData? If not, then you will probably just discard this and let farmOS create new Purchase logs with their own IDs. If it *is* referenced elsewhere, you'll have to figure out if/how to maintain that connection in farmOS to other related records.
- `purch_date` - This would map to the Purchase log timestamp.
- `type` - Ultimately, I think this *should* get mapped to an asset that is used for inventory tracking in farmOS. However, this feature is not available yet.  See related recent comments here: https://farmos.discourse.group/t/farmos-to-record-fertilizer-use/83/7 - In the future I would see each of these being assets in farmOS that are used to track inventory levels.
- `subtype` - This would probably be used in conjunction with `type` above to identify the "Material" asset that should be associated with the inventory adjustment (when that feature becomes available).
- `animal_group` - I assume this is the animal group that *will be* receiving the purchased material? In farmOS, once we have a proper "Materials" asset type then Purchase logs will increment the inventory of those assets, and then you would use an Input log to record when it is given to other assets.  So I don't think that should be directly tied to the purchase log itself, necessarily (because a purchase may get used on multiple assets, ultimately).  But for your purposes, you may consider just referencing animal/group assets directly on the purchase logs as a way of organizing things.  That might work for your purposes.
- `vendor` - Maps to the "seller" field in Purchase logs.
- `unit` - Maps to the "quantity units" field in Purchase logs.
- `purchased` - I think this would map to the "log owner" field in Purchase logs, if this represents "who bought it".
- `price_unit` - Maps to the "unit price" field in Purchase logs.
- `weight_unit` - We don't have a great place for this currently, but is something to think about with the inventory management features described above.
- `comments` - Maps to the "notes" field in Purchase logs.

I'm not sure how much AnimalData deals with tracking inventory of materials. You may need to wait for some of that stuff to come together before there's a truly 1:1 mapping available. But if it's not critical, then one option is to consider just dumping it into the notes field or title field, but it means you're losing the granularity of that data in the import.

**Comment by mstenta** [comment]
> Re: 1, taking a look at the farm_ledger module, I see we are not adding the importer mappings for purchase logs in the same way that we are for sale logs. That's easy enough to fix. That was an oversight when Purchase logs were added. I'll make a commit shortly...

Fixed: https://github.com/farmOS/farmOS/commit/da91c3d18f4e4e5d4c7586e788914f2604cfefdf

**Comment by megandalster** [comment]
This is great, thanks for your work on it.  I'll let you know how it goes after I've installed the patch here and done some importing -- maybe a couple of days.  Thanks again!

**Comment by megandalster** [comment]
Hi @mstenta,

Importing purchases is becoming a can of worms.  The above fields 'type' and 'sub_type' are really important for classifying feed purchase records from AnimalData.  I cannot just drop them.

Looking ahead at importing seed purchases, there are even more fields, as shown in the following export from FarmData:

crop | variety | year | source | catalogOrder | organic | catalogUnit | price | defUnit | unitsPerCatUnit | catUnitsOrdered | defUnitsOrdered | totalPrice | status | source1 | sdate1
BEAN | E-Z PICK | 2019 | JOHNNY'S | 242OG | OG | 5Msds | 33.55 | OUNCE | 1 | 1 | 1 | 33.5499992371 | ARRIVED |   | 2019-01-19
BEAN | JADE BUSH | 2019 | HIGH MOWING | 16G | OG | 1Msds | 6.55 | OUNCE | 10 | 5 | 50 | 32.7500009537 | ARRIVED |   | 2019-01-08
BEAN | RODCOR | 2019 | JOHNNY'S | 26G | OG | 1Msds | 10.04 | OUNCE | 1 | 2 | 2 | 20.0799999237 | ARRIVED |   | 2019-01-19
BEAN | ROYAL BURGUNDY | 2019 | JOHNNY'S | 2944G | OG | 5Msd | 31.15 | OUNCE | 1 | 1 | 1 | 31.1499996185 | ARRIVED |   | 2019-01-19
BEET | CHIOGGIA | 2019 | HIGH MOWING | 25G | OG | 1Msd | 4.4 | POUND | 5 | 5 | 25 | 22.0000004768 | ARRIVED |   | 2019-01-19
BEET | DETROIT DARK RED | 2019 | HIGH MOWING | 2280 | OG | 1 lb | 15.3 | POUND | 1 | 1 | 1 | 15.3000001907 | ARRIVED |   | 2019-01-19
BEET | EARLY WONDER | 2019 | HIGH MOWING | 2290 | OG | 1/4 lb | 7.1 | POUND | 0.25 | 1 | 0.25 | 7.0999999046 | ARRIVED |   | 2019-01-19

Here, the columns 'crop' and 'variety' are important names that cannot be dropped.  Somehow, the importers need to be modified to accommodate these fields.  They are like 'type' and 'sub_type' for ideitifying different feed materials.

Not sure how to proceed from here, but I hope this information helps you sort out how to import FarmData and AnimalData exports into FarmOS.

Meanwhile, I'm going to fiddle with adding a 'Materials' asset type and view, and see if somehow there's a way to name individual assets of this type, like 'Chicken Feed' or 'Chioggia Beet Seed'.  

Also, it seems as though 'Feed' and 'Seed' deserve to be their own asset types, and 'Materials' could be a catch-all for other purchased materials, like 'Bedding'.  Does that make sense?

 Let me know if/how I can be of further help with this issue.

**Comment by mstenta** [comment]
Yea this is going to take some work. FarmData and AnimalData provide features that farmOS does not. So there will be differences and each one needs to be figured out.

There are a spectrum of options, though.  What I described is how I see the long-term "ideal" solution, but that's not going to work for you in the short term.

Another potential approach would be to create a custom FarmData module for farmOS, which provides it's own database table(s) to extend logs with the specific fields that you need. This gives you a lot of flexibility.  You can essentially define your own schemas and link them to logs/assets/etc.  And as more generalized pieces take shape in farmOS, they can adopted/integrated into the module.

**Comment by megandalster** [comment]
Yes, creating a FarmData/AnimalData module for FarmOS seems to be the right approach.  Though I've been a Drupal user in the past, I'm not experienced building modules.  I do have a lot of PHP/MySQL experience with other software projects, so this might not be too tough to do.  

Do you have any recommendations for tutorials on getting started building a Drupal module?

**Comment by megandalster** [comment]
Also, should I build it for D7 or D8, or doesn't that matter too much?

**Comment by mstenta** [comment]
> so this might not be too tough to do.

Creating a module that adds database tables is very easy. And then from there you can begin learning how to tie those database tables into other aspects of the system (displaying their data alongside logs, exposing forms to edit them, etc).

A module is essentially just two file: `mymodule.info` (which describes the module to Drupal) and `mymodule.module` (which can be empty, but must exist).  If your module adds database tables, then a third file is necessary, `mymodule.install` (which contains code that will only be loaded during module installations, uninstallations, etc). Then, it's just a matter of implementing [hooks](https://www.drupal.org/docs/7/creating-custom-modules/understanding-the-hook-system-for-drupal-modules) (adding functions with specific naming patterns) in your `mymodule.module` and `mymodule.install` files.

Here is an example of a module that provides database tables (included in the [Examples](https://www.drupal.org/project/examples) module, which contains a ton of example modules/code for doing different things in Drupal): https://git.drupalcode.org/project/examples/blob/7.x-1.x/dbtng_example/dbtng_example.install#L46 (implementation of `hook_schema()` to define database tables provided by the module).

The Drupal API documentation is very helpful for understanding how to use particular hooks. Just Google "drupal hook_schema" and you'll find this: https://api.drupal.org/api/drupal/modules%21system%21system.api.php/function/hook_schema/7.x

That should get you up and running adding database tables, then it's just a matter of deciding what to do with them... you will have a better sense of your needs than I will, but happy to provide advice and guidance when I can!

> Also, should I build it for D7 or D8, or doesn't that matter too much?

I wish I could say D8, but farmOS just isn't available on D8 yet.  That does mean that you'll need to upgrade your module to D8 in the future when farmOS 2.x is ready.  But if you keep your custom module as simple as possible (and try to fit as much as you can into farmOS itself), then it will mean less work to upgrade when the time comes.

**Comment by mstenta** [comment]
Put your custom module(s) in `sites/all/modules/[modulename]`. The `sites` directory is preserved between farmOS version upgrades.  If you put anything in `profiles/farm/modules` (where core farmOS modules live) they will be lost when you upgrade.  (Of course, if you are maintaining your module's code in a Git repository, which you should, then it wouldn't be totally lost.)

**Comment by megandalster** [comment]
Okay, thanks for all this.  It will get me started.  

One more question -- can you project a time frame for completing the FarmOS inventory piece?  I don't want to do a lot of work that will be eventually subsumed by that.

Thanks again for all your help.  Hopefully I can contribute something that will be useful to others going forward.

**Comment by mstenta** [comment]
> can you project a time frame for completing the FarmOS inventory piece?

Unfortunately I can't.  I have to focus on sponsored development projects at the moment, and no one has expressed interest in sponsoring the remaining pieces of the inventory features yet.

**Comment by megandalster** [comment]
What would it cost for NPFI.org to sponsor this?

**Comment by mstenta** [comment]
> What would it cost for NPFI.org to sponsor this?

Good question... it's been a little while since I've looked at it.  The other consideration is timeline - I'm spread quite thin at the moment, so I might not be able to pick it up for a few months - which I imagine is too long for your purposes. I'm also a bit torn about whether or not it makes sense to do this before/after the D8 upgrade (really trying not to over complicate that upgrade any more than it already is).

However, it should be possible to design your module/schema with the future in mind, to minimize duplicate effort. You could probably create your own database tables of types and subtypes, which will eventually map to new assets. And a table to track increments/decrements to those in the same way that farmOS will via logs.

It might make sense to start by creating some Google Docs spreadsheets of the table schemas that exist in FarmData/AnimalData now, with columns to think through where they will need to map in farmOS... and from there figure out which features will need to be built to bridge the gap.  I really don't have a great understanding of FarmData/AnimalData, so something like that would help me to provide ideas/recommendations.

PS @megandalster - Thanks for the donation to the farmOS Open Collective!!

**Comment by megandalster** [comment]
Okay, thanks, this is good information.  I'm going to dive into something, maybe using a clone of your inventory module as a starting point for a contributed module.  That way, whatever I add on will be consistent with your initial inventory code.

I'm going to use the CSV exports from FarmData and AnimalData as profiles for any new database tables that I'll need -- all the reports from FD and AD have CSV export tabs, so the database table design should be relatively easy.  The tough part for me will be to learn the Drupal protocols for defining the module and its database interface.

I'll get back to you when I have something worth talking about :-)

Thanks again for your help and perspective on all of this!

**Comment by mstenta** [comment]
Sounds good! Good luck! Feel free to ping me with questions here or in the farmOS chat.

> The tough part for me will be to learn the Drupal protocols for defining the module and its database interface.

These two doc pages will help:

https://api.drupal.org/api/drupal/modules!system!system.api.php/function/hook_schema/7.x

https://www.drupal.org/node/146843

**Comment by megandalster** [comment]
Thanks for this.  Quick question.  I'm setting up a localhost apache server on my Mac and trying to run a clone of my FarmOS install here.  I've logged in successfully, but for some reason I'm getting "The website encountered an unexpected error.  Please try again later" whenever I try to access any of the admin pages.

I know this question is off-topic for this thread, but do you see a quick fix?  Many thanks!

**Comment by mstenta** [comment]
Hard to say - can you access /admin/reports/dblog? That shows internal Drupal error logs (which can also be written to /var/log/syslog if you enable the `syslog` module). They might also be in your Apache logs.

(Side note: one benefit of running farmOS via the official Docker image is you can be sure you have all system-level dependencies, Apache/PHP versions, etc. https://farmos.org/development/docker/)

Yea, I'd say start a new issue if you can't figure it out. I can tell you where to look to debug.

**Comment by megandalster** [comment]
Good, thanks!

**Comment by megandalster** [comment]
Hi @mstenta,

I've done some cleaning up of the AnimalData/FarmData CSV exports, and it is now clear that the key missing field in your purchase importers is just Asset Name.  

I used Asset Name aggressively when importing animals into FarmOS, and it worked well.  As you know, the import creates an asset for every line imported.  But every imported animal name (ear tag) is unique, and so its name can be used as a primary key.

This is not the case for seed or feed.  An import will have several instances of the same asset name, as in the following example:

Asset IDs | Areas | Notes | Categories | Asset name | Invoice number | Lot number | Quantity | Units | Unit price | Total price | Seller

  |   |   | Feed | GRAIN: ROASTED SOYBEANS |   |   | BAG: 50lb | 4 | 40 | 160 | JASON FRITZ
  |   |   | Feed | GRAIN: ROASTED SOYBEANS |   |   | BAG: 50lb | 1 | 40 | 40 | JASON FRITZ
  |   |   | Feed | GRAIN: SWEET FEED |   |   | BAG: 50lb | 2 | 9.79 | 19.58 | TRACTOR SUPPLY

This CSV has the same structure as your Purchase Log Importer, except for the added column Asset name.  

So if your importer were revised to expose Asset Name in this way, new assets named GRAIN: ROASTED SOYBEANS and GRAIN: SWEET FEED would be created.  But would the importer create 1 or 2 assets with the name GRAIN: ROASTED SOYBEANS?

I'm wondering if Asset Name were exposed in this way, I could build a primitive inventory module that kept track of amounts of different assets as they were purchased, applied, and sold?

At the very least, exposing Asset Name would allow me to import the cleaned-up FarmData and AnimalData data into FarmOS.  I've figured out how the rest of the FD/AD fields can be preserved in the import.

Another way to go with this might be for the importer to allow the CSV file to pre-fill the Asset ID column and not overwrite it when it is non-blank.

Thanks for your thoughts on this.

**Comment by megandalster** [comment]
PS I see no use for importing Asset IDs on Purchases (or Sales).  If this column were changed to import Asset Names instead, that would completely solve my problem.  That is, I could import all feed and seed purchases from FarmData and AnimalData into FarmOS!   Does this change make sense?

**Comment by megandalster** [comment]
Ideally, I'd like to import feed purchases using the structure shown below:

<img width="757" alt="Screen Shot 2019-09-15 at 4 31 04 PM" src="https://user-images.githubusercontent.com/4589017/64963497-51f8c600-d867-11e9-80da-40e02019a647.png">

That is, I created a new asset type Materials and a group FEED for organizing this import.  Does this make sense for capturing the Asset Names that go with the purchases?

**Comment by mstenta** [comment]
> it is now clear that the key missing field in your purchase importers is just Asset Name.

Well that is great news and very coincidental!  Because I will be working on adding that option soon!  It is something that another farm wants to sponsor as well.  Still figuring out exactly when... I'm pushing to finish some other things by the end of the month.  So if not before then, then soon after I think.

> So if your importer were revised to expose Asset Name in this way, new assets named GRAIN: ROASTED SOYBEANS and GRAIN: SWEET FEED would be created. But would the importer create 1 or 2 assets with the name GRAIN: ROASTED SOYBEANS?

Ah - but the new column option will not automatically create assets.  That would need to be done ahead of time by you in a separate import.

> I'm wondering if Asset Name were exposed in this way, I could build a primitive inventory module that kept track of amounts of different assets as they were purchased, applied, and sold?
> ...
> I created a new asset type Materials

Ah great ... well... it actually would be possible to use the existing inventory module on your new asset type with very minimal effort.  The only catch is: inventory adjustments to the Material assets will only be recorded as a number.  Not a number + unit.  But if that's OK for now, then this is all very possible already!

In order to do that, you will need to create a new module that provides the Material asset type.  Then, in that module, in `hook_default_farm_asset_type()`, all you have to do is define how inventory is handled on the asset.  For example, this is what the Animal asset definition looks like:

```
/**
 * Implements hook_default_farm_asset_type().
 */
function farm_livestock_default_farm_asset_type() {
  $items = array();
  $items['animal'] = entity_import('farm_asset_type', '{
    "type" : "animal",
    "label" : "Animal",
    "weight" : 0,
    "inventory" : { "enabled" : "1", "individual" : "1" },
    "rdf_mapping" : []
  }');
  return $items;
}
```

Note the `inventory` property, which sets `enabled` to `1` and `individual` to `1`.  I think for your purposes, you would set `individual` to `0`.  The individual property basically means: "if no inventory adjustment records exist for an asset, treat the asset as an individual (inventory = 1)" - this allows an Animal asset to be used for individual animals OR head-counts of animals.

**Comment by mstenta** [comment]
> it actually would be possible to use the existing inventory module on your new asset type with very minimal effort

I should clarify this statement. It will be minimal effort to *enable* inventory tracking on the Material asset type.

However, I think that being able to *import* inventory adjustments via CSV will require additional work.  I forget if that's possible right now, but I don't think so.  This would basically involve adding an additional importer mapping column, similar to the new "Asset Name" column I'll be adding.  So keep that in mind as you are planning your migration.

Something I haven't brought up yet, but is worth mentioning: Drupal has a module called Migrate API (https://drupal.org/project/migrate) which is specifically designed for one-time migrations of data from one system/database to Drupal.  It is similar to the Feeds module, which is what we use to build our CSV importers, but allows you to pull directly from a database rather than having the intermediate step of creating a CSV and then importing that. It is very robust and gives you a lot of control over where data ends up in the Drupal side. That said, it is also a lot more complicated to learn, and requires creating custom modules/code. So you might be better off sticking with the process you've been scoping out already (CSV import). I just thought I would mention it so you are aware of the other options.

One other small thing: please name your new asset type `material` instead of `materials` (the machine name specifically - the label doesn't really matter). That is the pattern we follow in farmOS, so when we create an official `material` asset type you will be able to migrate to it easily.

**Comment by megandalster** [comment]
This is great stuff, thank you!  Especially the part about eventually being able to import Asset Name on Purchase logs.  I'll wait until you add that, and in the meantime I'll start looking at the modifications you suggest for the inventory module.  I can create a couple of Purchase logs with Asset Names by hand just for testing purposes.

Also, I'm going to stick with the CSV import rather than use the Drupal migrate, since FarmData and AnimalData exports are needing some cleaning up before importing them into FarmOS.

**Comment by megandalster** [comment]
Hi @mstenta, 

I have just created a bare-bones Material module and imported all the different FEED and BEDDING Material groups that were purchased by the Dickinson Farm and exported from FarmData/AnimalData.  I'm now ready to import the FEED and BEDDING purchases themselves as soon as you have had a chance to add Asset Name to the FarmOS Purchase Log importer.  No pressure :-)

You can take a look at all of this by going to https://npfi.org/farmdata2/user and logging in as admin user "mstenta" (password "mstenta").  That will show you where I am overall with the FarmData/AnimalData migration to the FarmOS platform.  Feel free to be critical ... this is all new to me!

The new Material module is basically a clone of the Compost module.  I'll need to add a couple of new fields for managing inventory -- Amount and Units -- at least, so that eventually each asset will be able to keep a running inventory of itself.  I've got more learning to do about Drupal modules, hooks, and overall FarmOS architecture before that happens...  Once that Module is done and tested, I will be happy to provide it as a Contributed Module to FarmOS, if you think it can integrate well with the Inventory module.

Thank you for your feedback...

**Comment by megandalster** [comment]
FYI, I just added 2 new Purchase logs by hand, both for GRAIN: ROASTED SOYBEANS, and their inventory adjustments worked perfectly.  Thanks for the tip on the hook!

Now I'm back to learning about adding and exposing more fields for Material assets...

**Comment by mstenta** [comment]
> I have just created a bare-bones Material module and imported all the different FEED and BEDDING Material groups that were purchased by the Dickinson Farm and exported from FarmData/AnimalData.

That's great!!!

> as soon as you have had a chance to add Asset Name to the FarmOS Purchase Log importer. No pressure

It's still on my list!  Hopefully soon... ;-)

> FYI, I just added 2 new Purchase logs by hand, both for GRAIN: ROASTED SOYBEANS, and their inventory adjustments worked perfectly. Thanks for the tip on the hook!

Great!!

>  I'll need to add a couple of new fields for managing inventory -- Amount and Units -- at least, so that eventually each asset will be able to keep a running inventory of itself.

We have an open issue for this here: https://www.drupal.org/project/farm/issues/2916551

Last time I was in that code, I was thinking that the best course of action might be to integrate the inventory fields into the farmOS Quantity Measurements field, because that already has units, labels, etc.  It's not a quick task, unfortunately.  And part of me wants to wait until we upgrade to Drupal 8 too - since there will be some significant refactoring involved (https://www.drupal.org/project/farm/issues/2991378 and https://www.drupal.org/project/farm/issues/2361735).

> Now I'm back to learning about adding and exposing more fields for Material assets...

Quick tip: all farmOS fields can be edited with the Field UI module, and exported to code with the Features module.  You rarely have to write/edit code for fields directly, unless you're doing something complicated.

**Comment by mstenta** [comment]
> It's still on my list! Hopefully soon... ;-)

I've started looking into this. I'm tracking my progress in this issue: https://www.drupal.org/project/farm/issues/3086611

**Comment by mstenta** [comment]
I've added support for a new "Asset names" column. You should be able to pull the latest `7.x-1.x` commit and it will "just work". Let me know if you have any trouble with it!

**Comment by mstenta** [comment]
See this commit to farmOS.org for documentation of the new "Asset names" column, as well as known issues/limitations: https://github.com/farmOS/farmOS.org/commit/c8581d73c5754d16f6338c750e48096f8b19f0c3

This has not been published on farmOS.org yet - it will be when farmOS 7.x-1.3 is released.

**Comment by megandalster** [comment]
Great!  Thanks for all your work on this.  I'm still on the road, and will get back to it next week.

**Comment by megandalster** [comment]
Okay, I have successfully uploaded feed and bedding purchase logs to farmdata2, using your new importers that include Asset Name.  After I did a bunch of data cleanup on the FarmData end, your importers worked like a charm!

Next step is to import animal medical records from AnimalData into FarmData2.  I'll keep you posted on this.

**Comment by mstenta** [comment]
Great!!!

**Comment by megandalster** [comment]
Hi Again,

I'm now importing medical records from AnimalData to our FarmOS instance at the Dickinson Farm.  I'm looking at your CSV importer for medical logs, and several important fields are missing -- I cannot drop this medical data exported from AnimalData.  Here's a quick summary of the fields that I need to capture in a FarmOS medical log import (the Date, Asset names, and Notes fields are already there -- the rest are not):

Date (care_date)
Asset names (Animal ID)
Notes (Medical Condition)
Symptoms
Weight
Vet
Meds -- type, units, amount, dosage
Medical Outcome

Here is some sample CSV data that I exported from AnimalData, just to give you a feel for this:

id | care_date | animal_id | reason | symptoms | temperature | care | weight | vet | contact | assistants | comments | userid | medication | units | units_given | dosage
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
7 | 2016-08-12 | C-16-93 | INFECTED UMBILICAL STUMP | Infected stump at 1 week of age.  Swelling and fever (107) | 107 | N/A | 100 | Rhodes | PHONE | Kenze | Meds administered under vet's advice.  Calf snapped out of it and rallied, fine shape | steimanm | BANAMINE | mg | 1 | 1 CC PER 100LB
7 | 2016-08-12 | C-16-93 | INFECTED UMBILICAL STUMP | Infected stump at 1 week of age.  Swelling and fever (107) | 107 | N/A | 100 | Rhodes | PHONE | Kenze | Meds administered under vet's advice.  Calf snapped out of it and rallied, fine shape | steimanm | BANAMINE | mg | 1 | 1 CC PER 100LB
8 | 2016-08-12 | C-11-83 | MASTITIS | Calf born late in heat of summer.  Bagged up big before calf born.  Right rear quarter distended, hard and hot | 103 | Separated from herd until well, kept with calf.  Milked udder and massaged with mint balm as often as possible. Meds administered under vet's advice over 2 weeks time. | 1400 | Rhodes | EMAIL | Kenze | Cow rallied, calf fine. | steimanm | BANAMINE | mg | 115 | 1 CC PER 100LB
13 | 2017-05-25 | S-10-JB | DOWN AND OUT | 104 temp. Lumpy under. | 104 | Her right side teet was too big for the lambs to nurse on so we milked here for a few minutes to help the swelling go down. There were no visible solids in her milk. When we were done milking both lambs were able to nurse at the same time. | 130 | Conor / rhodes | PRESENT | Lizzie and Danielle | Temp down to 102.5 day of banamine . 6/2- 103.1 temp | wilforde | BANAMINE | cc | 1.5 | 1 CC PER 100LB

I think when Matt Steiman and Tim Wahls designed AnimalData, they thought of the log entry as being the place to record a "Vet Care" event occurring on a certain date.  Your idea of a medical log may or may not reflect the same thought -- I don't know.

Let me know what you think of this, especially 1) how to accommodate the additional fields and 2) about their general usability for FarmOS medical logs outside of the Dickinson Farm.

Many thanks, as always!

**Comment by megandalster** [comment]
Hmmm... maybe each of the above lines really depicts 3 separate logs in FarmOS -- a medical condition, a medical treatment, and a medical outcome (defining 3 categories in FarmOS).  The 3 could then be relinked together by sharing the same Log Name in FarmOS.

This approach doesn't get rid of the need for adding more fields to a medical log entry, but maybe not so many.

Just musing.... Does this make sense to you???

**Comment by mstenta** [comment]
Hmm all good thoughts! When we added the Medical logs originally they were not (and still aren't) very detailed in the information they collect. They basically just have a title and a notes field.

I think some of the fields you describe would be good additions. Notably there should be a "Quantity" field (same as on other log types). That alone would cover a number of your columns including:

- temperature
- medication
- units
- units_given
- dosage

> maybe each of the above lines really depicts 3 separate logs in FarmOS --  a medical condition, a medical treatment, and a medical outcome

That's a good thought as well...

- "Medical condition" might just be an "Observation" log
- "Medical treatment" could use the "Medical" log (with the addition of a Quantity field on Medical logs for recording medicine and dosage info)
- "Medical outcome" might also be an "Observation"

It makes sense because in practice you won't have all three of those things happening at the same time - so they aren't technically the same event (log).  But you make a good point that they should be linked together. I have considered in the past adding a general "Related Logs" field, which would let you reference other logs from a log. Haven't had a real need for that yet, but I think it could be useful.

**Comment by mstenta** [comment]
"Vet" would be another good field addition (reminds me of "Laboratory" field in the Soil Test log).

I'm not sure if "Symptoms", "Care", and "Notes" need to be three separate fields though. They may just be the Notes fields in multiple separate logs (separate observations and medical treatments).

**Comment by mstenta** [comment]
I created two specific feature requests for adding "Quantity" and "Vet" fields to Medical logs:

- https://www.drupal.org/project/farm/issues/3092756
- https://www.drupal.org/project/farm/issues/3092757

They won't be too hard to add, so I'll try to do that soon.

**Comment by megandalster** [comment]
Great... I especially like your thoughts about how to define the 3 separate logs and repurposing the Notes field for each one.  

Thanks for the feature requests... I'm glad that won't be too hard on your end.

PS, there's another field, Weight, that should be added to the medical log, since many of the prescribed dosages are based on the animal's weight at the time of treatment.

**Comment by mstenta** [comment]
> PS, there's another field, Weight, that should be added to the medical log, since many of the prescribed dosages are based on the animal's weight at the time of treatment.

Two thoughts on that:

1. farmOS provides the ability to track animal weights, so if you use that you could potentially load the animal's weight from the system.
2. Adding a "Quantity" field to the Medical log will allow you to add a Weight measurement in addition to other quantity measurements, if that's still useful for recording.

That brings another consideration to mind: The CSV importers are currently limited to importing a single quantity measurement per log (even though logs support unlimited measurements). So this might make it difficult to import everything you need in a Medical log (unless that stuff is broken up into other observation logs).

**Comment by mstenta** [comment]
> The CSV importers are currently limited

Side note: I wonder if you might want try the farmOS.py Python library instead of the CSV importers for more advanced things like this. There are no limitations to what can be imported/exported with that.

https://github.com/farmOS/farmOS.py

**Comment by megandalster** [comment]
Okay thanks, I'll take a look at that down the road.  (I'm having too much fun right now importing inputs and seedings :-)

**Comment by megandalster** [comment]
I've just imported 6000+ harvests from FarmData to FarmData2, our new instance of FarmOS at the Dickinson Farm.  Worked _almost_ like a charm -- log in and take a look if you get a chance.

Here's the _almost_ part -- FarmData kept records of labor "hours" for the person assigned to each harvest.  This information is important to Matt when he creates new assignments for future harvests and other logged activities.  I understand that that might be best done at the FarmOS-client level, but it doesn't help at bulk import time.

Could an "hours" field be added to the Assignment part of the Log profile in FarmOS?

Thanks for considering this!

**Comment by megandalster** [comment]
The CSV importers for Seeding Log and Transplanting Log are both missing an Area column (this field is called "Movement To" in their respective pages.  I have 500+ Logs of Seedings and 600+ Logs of Transplantings to import from Dickinson's current FarmData software, and the Area data is essential.

Can you add an Area column to each of these two importers?  I'm dead in the water until that can be done.

PS this is a far more critical request than in my previous note, which can wait for now.

Many thanks!

**Comment by megandalster** [comment]
Hi @mstenta 

I have just patched the feeds importers so that Seeding and Transplanting logs will bulk-import Areas (called Movements) with their individual seedings and transplantings that are exported from FarmData.  I will do the importing tomorrow and will then commit commit the patch to my instance of the code once I'm confident that the patch is solid.  

I'll keep you posted -- this may be a patch that other FarmOS users will want as well.

**Comment by megandalster** [comment]
Help!  I'm having a tough time getting the Seeding importer to work after adding the Movement field to its CSV template.  Here's what I've changed in the farm_crop.module, where the importers for seedings (and transplantings) are located:

In the function farm_crop_feeds_importer_default_alter I added the "Movement to" mapping:
<img width="620" alt="Screen Shot 2019-11-25 at 12 14 45 PM" src="https://user-images.githubusercontent.com/4589017/69562781-de1d1d00-0f7d-11ea-975d-bc1c00060fba.png">

In the function farm_crop_feeds_tamper_default_alter, I added the code:
<img width="695" alt="Screen Shot 2019-11-25 at 12 15 07 PM" src="https://user-images.githubusercontent.com/4589017/69562859-0ad13480-0f7e-11ea-8a62-f115c1ba05db.png">

So the CSV file now has the folllowing contents:
Done | Date | Log name | Asset IDs | Asset names | Notes | Categories | Quantity measure | Quantity value | Quantity unit | Quantity label | Source/supplier | Movement to
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  | 2016-03-15 |   |   | Celeriac | Seed Code: CC-16-1-OG (DIAMANTE) - 400 seeds;Seed Code: CC-15-1-OG (DIAMANTE) - 400 seeds | Seeding: Tray | Count | 3.75 | Flats | 128 seeds/flat |   | Seeding Bench

And the resulting import works, but drops the Movement to field, as you can see:
<img width="1192" alt="Screen Shot 2019-11-25 at 12 13 27 PM" src="https://user-images.githubusercontent.com/4589017/69563008-5257c080-0f7e-11ea-906c-09548540dfd1.png">

I have the same problem with importing Transplantings, but I think the solution for that will be similar to solving this one.

i haven't altered the farm_crop.info file at all, though I'm still unclear about the difference between the "movement" field vs the "move_to" field in the farm_crop module.

Many thanks for any light you can shed on what I'm doing wrong.

**Comment by mstenta** [comment]
Sorry for losing track of this @megandalster ! It got buried in my inbox, then the holidays hit!

> FarmData kept records of labor "hours" for the person assigned to each harvest

We have a plan to add time tracking in the future. See https://www.drupal.org/project/farm/issues/2402943

I'm not sure what to suggest in the meantime. The quantity field could be used to store this info, but the CSV importers only support a single quantity measurement right now, and you're already using that for your harvest info. Also, when we do add official time tracking support it will most likely be in its own storage field - not quantity measurements, so you would end up with legacy data in one place and new data in another place. Maybe that's not a bad thing, though, as long as you're aware of it (and you could potentially automate the migration of legacy data to the new field, too, as long as you were consistent in the usage of it).

> The CSV importers for Seeding Log and Transplanting Log are both missing an Area column (this field is called "Movement To" in their respective pages.

Yes, unfortunately the CSV importers do not currently support movements. :-(

>  I'm having a tough time getting the Seeding importer to work after adding the Movement field to its CSV template.
> ...
> In the function farm_crop_feeds_importer_default_alter I added the "Movement to" mapping:

Unfortunately, it is not quite as simple as what you attempted. The "Movement" fields are not normal fields. They are Field Collections, which means they are actually a separate entity with sub-fields on it ("Movement to" and "Geometry").

In order to be add a "Movement to" column to the CSV importers, it will require writing a custom Feeds mapping target.

https://www.drupal.org/docs/7/modules/feeds/the-developers-guide-to-feeds/the-developers-guide-to-feeds#s-mapping-api

This mapping target would need to take the area name that is included in the CSV, use the `farm_term()` function to get the associated taxonomy term (or create a new one), then create a new Movement field collection entity linked to the log that is being created.

FWIW, I want to phase-out our use of Field Collections in the future, because it makes things like this more complicated: https://www.drupal.org/project/farm/issues/2991374

However, that probably won't happen in the 1.x branch of farmOS. We are turning our focus to the 2.x branch now, so that will be an opportunity to rethink it.

If you would like to try writing a feeds mapping target for movement logs, please feel free! It is a bit of a deep dive into a few different Drupal architecture worlds though, so I won't say it's "easy". If you do, let's open a dedicated feature request for that, apart from this one (which is getting very long!)

These are some of the "devil in the details" I had a feeling we would run into as we started to get deeper into the FarmData->farmOS migration. It might make sense to create a dedicated meta issue to keep track of requirements like this. I wish I could dedicate more time myself to help

**Comment by mstenta** [comment]
FWIW, I see there is a module that claims to "Provide feeds integration for field collection module": https://www.drupal.org/project/field_collection_feeds

It is marked as "alpha", however, and has not been touched in 4 years, so I don't know what its status is. Might be worth a try with some test data to see if it would work for this. I don't think we would include this as a dependency with farmOS, however, due to it's status (and the likely move away from Field Collections in Drupal 8), so it would be a custom modification that you make only to your own farmOS instance.

**Comment by megandalster** [comment]
Thanks, @mstenta, no worries about the delay... life happens!  FYI, I'm on my way to Dickinson College next week and will be teaching there this spring semester.  

Meanwhile, I will take up your suggestion on developing a custom Feeds Mapping Target.  We really need it to complete the migration of data over from FarmData.   I'll keep you posted on this one.

I also like your idea of starting a dedicated meta issue on FarmData->FarmOS migration.  I'm almost there with the migration of the Dickinson Farm data, especially if I can get the Feeds Mapping Target done.  Once that is done, I will document my experience with examples that others can replicate -- migrating other FarmData instances should be pretty straightforward if a good roadmap is created.

Thank you for all your help and sticking with this tough issue.  You've taught me a lot along the way!!!

**Comment by mstenta** [comment]
FYI @megandalster - I added Quantity and Veterinarian fields to Medical logs:

- https://www.drupal.org/project/farm/issues/3092756
- https://www.drupal.org/project/farm/issues/3092757

Hope you're doing OK these days!

**Comment by megandalster** [comment]
Great, thanks for this!  I'm doing fine, thanks for asking.

I just got back from a challenging semester at Dickinson College -- lots of stuff to talk about, but not on this thread :-)

PS I like your Docker development bundle... much easier than MAMP!

**Comment by megandalster** [comment]
Hi @mstenta,

I just finished adding a custom Feeds Mapping Target for the "Movement to" field collection to the farm_crop.module's feeds importer.  It works well, and gets us over a big hump importing FarmData Seeding and Transplanting Logs into our FarmData2 variant of farmOS.  Your guidance was perfect, thank you!

I'm attaching that piece of the modified farm_crop.module that I coded to get this done.  This new code is probably not perfect, but it does work for our CSV imports.  Also, if you think this should get retested, refined, and passed on as a contribution for the larger farmOS community, I'm all for it.  

Cheers, and thanks again for your help with this!

[farm_crop.modulenewcode.pdf](https://github.com/farmOS/farmOS/files/4932038/farm_crop.modulenewcode.pdf)

**Comment by mstenta** [comment]
I'm cleaning up the GitHub issue queue, so I'm going to go ahead and close this @megandalster. We're encouraging everyone to start using the farmOS forum for discussions like these going forward: https://farmOS.discourse.group

Speaking of - I'd love to hear how things are going with FarmData/Dickinson! And especially with farmOS 2.x in alpha now (and approaching beta) - it would be great to start preparing for your eventual transition. Happy to help think that through with you in the forum. Hope all's well!
